Automattic/wp-calypso
# [PBD Interim fix] Update getExitCheckoutUrl to use the upgradeIntent for plugins and themes

This change allows the checkout to use a `redirect_to` parameter when a user empties their cart. The `getExitCheckoutUrl` function will iterate over the `redirect_to` trying to find the deepest buried redirect, and will use that as the redirect location if a user has a valid upgrade intent. It will not follow the redirect if it is to an external URL.

When the changes to the bannerUrls for upgrading to the business plan is merged in https://github.com/Automattic/wp-calypso/pull/38203, this PR will 'switch on' the behaviour where a user who goes from the `/plugins` page to upgrade their site, and then cancels on the checkout, will be redirected back to the plugins page. The same with themes, or any individual plugin. That change is _not_ part of this PR.

#### Changes proposed in this Pull Request

* Add `upgradeIntent` and `redirectTo` to `getExitCheckoutUrl` so that a request to the checkout where a user has nothing in the cart (e.g. after they click the trash icon on their cart items) redirects to the provided `redirectTo`
* Add constants for each of the valid upgrade intents and a function for validating a provided intent against them
* Add function to parse a `redirect_to` string that includes nested redirects that may or may not be URI encoded
* Add test coverage for the new functions

When the upgrade banners are eventually updated with intent and redirect URLs the intended behaviour will look like:

![plugin-checkout-redirect-small](https://user-images.githubusercontent.com/14988353/70302621-a782b480-1851-11ea-9813-6f917f3c8c04.gif)

#### Testing instructions



##### Test for regressions

In the below examples replace `site-name.blog` with the domain for your test site.

* Ensure that if you sign up via `/start` and reach the checkout on a paid plan, that clicking the trashcan icon in checkout redirects you to `/plans`
* From an existing site, go to add a domain via `/domains/add/site-name.blog` (you can be on a free tier and add a domain with a plan by clicking Upgrade next to a domain). Click on the trashcan icon in checkout, and you should be redirected to `/domains/add/site-name.blog`
* Go to purchase a premium theme `/themes/premium/site-name.blog` (click Show All Theme, then on the ellipsis next to a theme, select Purchase). Click on the trashcan icon in checkout, and you should be redirected to `/themes/site-name.blog`
* Create a Jetpack site via https://jurassic.ninja/ — after the site is created, set up Jetpack and connect it to WordPress.com. When you get to `wordpress.com` URLs, replace the URL with `calypso.localhost:3000/` and from the plans step select a paid plan. From the checkout screen, click the trashcan icon and you should be redirected to `/plans/my-site-name.jurassic.ninja`. From this screen go to upgrade the plan again with any of the upgrade buttons. From the checkout, click the trashcan icon, and again you should be redirected back to `/plans/my-site-name.jurassic.ninja`

##### Test this change

In the below examples replace `site-name.blog` with the domain for your test site.

* From an existing site go to `/plugins/site-name.blog` and click the upgrade nudge to upgrade to the Business plan. From the checkout, update the URL in your browser to be the same page, but include the query string `?intent=install_plugin&redirect_to=/plugins/site-name.blog` (remember to replace `site-name.blog` with the domain of your site) and reload the page.
* Click the trashcan icon and you should be redirected to `/plugins/site-name.blog`
* Repeat the above steps but use the query string `?intent=install_plugin&redirect_to=%2Fplugins%2Fsite-name.blog%3Fredirect_to%3D%2Fme` and reload the page.
* Click the trashcan icon and you should be redirected to `/me`

Fixes #38144
